### PR TITLE
Storage Iteration functions for accounts

### PR DESCRIPTION
**Pull request by dsainati1**
Part of https://github.com/onflow/cadence/issues/208
## Description

This adds three new functions to the `AuthAccount` object `forEachPublic`, `forEachPrivate` and `forEachStored` (as well as just `forEachPublic` to `PublicAccount`) that iterate over all the paths (and the types of the objects stored there) in the respective domains of the account. 

Each of these functions takes a function parameter of type `(X, Type): Bool` (where `X` is the type of the path being iterated over by the function in question (`PublicPath`, `PrivatePath` and `StoragePath` respectively)), which is applied to every element stored in that account under the domain in question. The first parameter holds the path of the element, and the second holds its runtime type.

These paths and types are loaded from storage one by one, to avoid loading the entirety of storage into memory in the case of large accounts. 

The specific order in which the objects are iterated over is undefined, as is the behavior when a path is added or removed from storage, whether by linking, unlinking, saving or loading. The behavior is, however, deterministic; we make no guarantees about what we do but we do the same thing every time. 
______

- [X] Targeted PR against `master` branch
- [X] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [X] Code follows the [standards mentioned here](https://github.com/onflow/cadence/blob/master/CONTRIBUTING.md#styleguides)
- [X] Updated relevant documentation 
- [X] Re-reviewed `Files changed` in the Github PR explorer
- [X] Added appropriate labels

### CODE REVIEW

File: docs/language/accounts.mdx [hunk 1]
```diff
@@ -121,6 +124,11 @@ to the `prepare` phase of the transaction.
       fun getLinkTarget(_ path: CapabilityPath): Path?
       fun unlink(_ path: CapabilityPath)
 
+      // Storage iteration  
+      fun forEachPublic(_ function: ((PublicPath, Type): Bool)): Void
```
Comment (dsainati1) [inline docs/language/accounts.mdx:128]: If we don't make any guarantees about the order of iteration, is it actually useful to be able to terminate iteration using this boolean return value? What use cases would there be for wanting to get, say, 3 random NFTs out of storage without caring specifically which ones they are?

Comment (bluesign) [reply docs/language/accounts.mdx:128]: Searching for specific `Type` can be some use case. Normally people will have one Collection Type in storage.

**Comment by bluesign** [comment]
> These paths and types are loaded from storage one by one.

Why do we load full value here ? Shouldn't we load type info only ? This can cause big usability problems.

**Comment by dsainati1** [comment]
We do load only the runtime type info here, the actual stored value itself is not loaded.

File: runtime/interpreter/interpreter.go [hunk 2]
```diff
@@ -3673,6 +3673,50 @@ func (interpreter *Interpreter) allAccountPaths(addressValue AddressValue, getLo
...
+
+			for key, value := storageIterator.Next(); key != "" && value != nil; key, value = storageIterator.Next() {
+				pathValue := NewPathValue(invocation.Interpreter, domain, key)
+				runtimeType := NewTypeValue(invocation.Interpreter, value.StaticType(invocation.Interpreter))
+
+				subInvocation := NewInvocation(
+					interpreter,
+					nil,
+					[]Value{pathValue, runtimeType},
+					[]sema.Type{pathType, sema.MetaType},
```
Comment (SupunS) [inline runtime/interpreter/interpreter.go:3699]: Can `[]sema.Type{pathType, sema.MetaType}` be extracted out and re-use across iterations? I guess it is not updated in downstream codes?

File: docs/language/accounts.mdx (hunk 1 above)
Comment (SupunS) [inline docs/language/accounts.mdx:128]: I understand that only the `Type` is made available here to avoid loading the value. But in the implementation, the value is already loaded anyway (i.e: `storageIterator.Next()` returns the value), so wouldn't it be better to just make the value available here as well? I think loading a value doesn't load the entire value graph (i.e: lazy loading). So it should be fine?

Comment (dsainati1) [reply docs/language/accounts.mdx:128]: The problem with making the value available is that if it is a resource, it needs to be removed and then resaved in order to avoid implicitly duplicating it (which can also re-order things in storage). By just providing the type and then making the developer handle loading resources explicitly we avoid implicit duplication.

Comment (SupunS) [reply docs/language/accounts.mdx]: ah ok, that makes sense!

**Review by fxamacker — COMMENTED** [review]
Nice work!  I focused on atree and storage related code.  I left a couple comments related to handling an edge case where `storageMap` can be nil.

File: runtime/tests/interpreter/account_test.go [hunk 3]
```diff
@@ -1942,157 +1945,543 @@ func TestInterpretAccount_iteration(t *testing.T) {
...
 
-		value, err = inter.Invoke("getPublicPaths")
+			fun test(): Int {
+				account.save(S(value: 2), to: /storage/foo)
+				account.save("", to: /storage/bar)
+				account.link<&S>(/public/a, target:/storage/foo)
+				account.link<&String>(/public/b, target:/storage/bar)
+				account.link<&S>(/public/c, target:/storage/foo)
+				account.link<&S>(/public/d, target:/storage/foo)
+				account.link<&String>(/public/e, target:/storage/bar)
```
[the remark this answers is not in the document]
Comment (dsainati1) [reply runtime/tests/interpreter/account_test.go:2125]: > Also, do we need to test path inside iteration functions?

Do you mean do we need to test if the path is correct? Some of these tests use `borrow` on the path value, which would only work if the path is accurate.

**Review by fxamacker — APPROVED** [review]
Great work! :+1:  I focused on atree and storage aspect.

File: runtime/interpreter/interpreter.go [hunk 4]
```diff
@@ -3671,6 +3671,57 @@ func (interpreter *Interpreter) storageAccountPaths(addressValue AddressValue, g
 	return interpreter.accountPaths(addressValue, getLocationRange, common.PathDomainStorage, PrimitiveStaticTypeStoragePath)
 }
 
+func (interpreter *Interpreter) iterOverStorageDomain(addressValue AddressValue, domain common.PathDomain, pathType sema.Type) *HostFunctionValue {
```
Comment (turbolent) [inline runtime/interpreter/interpreter.go:3674]: Maybe improve naming here: starting the name of the function with a verb makes it sound like it performs the iteration, but it does not, it only returns a function. A common naming scheme is e.g.

```suggestion
func (interpreter *Interpreter) newStorageIterationFunction(addressValue AddressValue, domain common.PathDomain, pathType sema.Type) *HostFunctionValue {
```

File: runtime/sema/authaccount_type.go [hunk 5]
```diff
@@ -222,6 +243,45 @@ const authAccountTypePrivatePathsFieldDocString = `
 All the private paths of an account
 `
 
+const authAccountForEachPublicDocString = `
+Iterate over all the public paths of an account.
+
+Takes two arguments: the first is the path (/domain/key) of the stored object, and the second is the runtime type of that object.
```
Comment (turbolent) [inline runtime/sema/authaccount_type.go:249]: The iteration functions accept one argument, the iteration function. The iteration function then is passed two arguments.

File: runtime/tests/checker/account_test.go [hunk 6]
```diff
@@ -1644,3 +1644,242 @@ func TestCheckAccountPaths(t *testing.T) {
...
+			correctType string
+		}{
+			{name: "forEachPublic", correctType: "PublicPath"},
+			{name: "forEachPrivate", correctType: "PrivatePath"},
+			{name: "forEachStored", correctType: "StoragePath"},
+		}
+
+		for _, pair := range nameTypePairs {
+			t.Run("basic", func(t *testing.T) {
+				t.Parallel()
```
Comment (turbolent) [inline runtime/tests/checker/account_test.go:1778]: Careful: Using `t.Parallel` will pause the closure, then restart it. By the time it's restarted, the `pair` variable of the outer for-loop has been rebound. This basically results in totally random behaviour, in the worst case only  the last element of `nameTypePairs` is tested 3 times.

You can fix this by factoring the body of the loop into a helper function and then calling it, e.g. (pseudo-code)

```go
test := func(pair /* ... */) { 
	t.Run(fmt.Sprintf("basic: %s", pair.name), func(t *testing.T) {
		t.Parallel()
		// ...
	})

	// ... 
}

for _, pair := range nameTypePairs {
	test(pair)
}
```

This rebinds `pair`.

Also, the current loop runs a sub-test "basic" for each function, maybe use the pair's name in it, otherwise you get generated subtest names like `basic #0`, `basic #1`, etc.

**Review by turbolent — COMMENTED** [review]
Nice!

File: runtime/interpreter/account.go [hunk 7]
```diff
@@ -86,6 +86,15 @@ func NewAuthAccountValue(
 		sema.AuthAccountStoragePathsField: func(inter *Interpreter, getLocationRange func() LocationRange) Value {
 			return inter.storageAccountPaths(address, getLocationRange)
 		},
+		sema.AuthAccountForEachPublicField: func(inter *Interpreter, _ func() LocationRange) Value {
+			return inter.iterOverStorageDomain(address, common.PathDomainPublic, sema.PublicPathType)
```
Comment (robert-e-davidson3) [inline runtime/interpreter/account.go:90]: The `Value` here is the `*HostValueFunction`. I'm not sure what a `HostValueFunction` is but should it be invoked here?

Comment (dsainati1) [reply runtime/interpreter/account.go:90]: No. These define a method on the account object. They are invoked when the method is called, not when it is defined.

**Review by turbolent — APPROVED** [review]
Nice work! 👍
